processing/src/main/java/io/druid/query/GroupByMergedQueryRunner.java

```diff
@@ -79,7 +81,108 @@ public GroupByParallelQueryRunner(
   }
 
   @Override
-  public Sequence<T> run(final Query<T> queryParam, final Map<String, Object> responseContext)
+  public Sequence<T> run(final Query<T> query, final Map<String, Object> responseContext)
+  {
+    boolean isSingleThreaded = query.getContextValue(CTX_KEY_IS_SINGLE_THREADED, configSupplier.get().isSingleThreaded());
+    if (isSingleThreaded) {
+      return runSingleThreaded(query, responseContext);
+    } else {
+      return runParallel(query, responseContext);
+    }
+  }
+
+  private Sequence<T> runSingleThreaded(final Query<T> queryParam, final Map<String, Object> responseContext)
+  {
+    return new ConcatQueryRunner<>(
+        Sequences.map(
+            Sequences.simple(queryables),
+            new Function<QueryRunner<T>, QueryRunner<T>>()
+            {
+              @Override
+              public QueryRunner<T> apply(final QueryRunner<T> input)
+              {
+                return new QueryRunner<T>()
+                {
+                  @Override
+                  public Sequence<T> run(final Query<T> query, final Map<String, Object> responseContext)
+                  {
+                    final Pair<IncrementalIndex, Accumulator<IncrementalIndex, T>> indexAccumulatorPair = GroupByQueryHelper
```

Comment:
@xvrl change I'm going to make is going to be a bit more than refactoring. The reason for separate PR proposal was because I wanted to remove creation of multiple indexes in single threaded case like is done for the parallel runner which changes the implementation of single threaded case really and not just moving code around for refactoring.

That said, I can see your point and will update this PR itself.